
# <img src="/doc/logo.svg" width="240" height="108"/>

YAX is a comprehensive software toolkit for analysis of molecular quantitative trait loci (xQTLs), such as mRNA expression or methylation levels. It performs cis and trans analysis, single- and multiple-variant analysis, and single-study and meta-analysis. 

We have conducted extensive benchmarking and simulation experiments to ensure that YAX is accurate, efficient, and scalable for small and large sample sizes; [**see benchmarking**](/doc/benchmarking/) for further details. 

To install YAX from source or download precompiled binaries, [**see installation guide below**](#installation).  Notable features include

**Fast, scalable linear mixed model for xQTL mapping**
 - Highly optimized linear mixed model (LMM) framework for tens of thousands of traits. This can be used to account for cryptic or known familial relatedness in cis and trans xQTL analysis. 
 - Optimized ordinary least squares (OLS) framework for cis and trans xQTL analysis with no related individuals.

**Multiple-variant and omnibus xQTL analysis**
 - Fast, rigorous analytic p-values for cis-region omnibus tests. 
 - Variable selection and conditional analysis procedures to identify multiple association signals for a single molecular trait.
 
**Flexible multiple-variant xQTL meta-analysis framework** 
 - Compact storage formats for xQTL summary statistics, including study-specific LD information. 
 - Single and multiple-variant xQTL meta-analysis procedures.
 - R interface for joint and conditional analysis, Bayesian finemapping, aggregation tests, and penalized regression from xQTL summary data files across one or multiple studies.
 
## Analysis modes and documentation
 - [**cis-xQTL analysis.**](/doc/mode_cis/) `./yax cis {options}`
	 - Analyze xQTL associations within a genomic window of each molecular trait.  For example, analyze variant associations within 1 megabase (Mb) of each gene's transcription start site (TSS).  
 - [**trans-xQTL analysis.**](/doc/mode_trans/)  `./yax trans {options}`
	 - Analyze xQTL genome-wide associations for each molecular trait, optionally excluding variants within the cis region.   
 - [**xQTL meta-analysis.**](/doc/mode_meta/)  `./yax meta {options}`
	 -  Meta-analyze xQTL association summary statistics across one or multiple studies.   
 - [**xQTL summary data storage.**](/doc/mode_store/)  `./yax store {options}`
	 -  Store genotypic variance-covariance (LD) data matrices for data-sharing or meta-analysis. These files enable accurate multiple-variant analysis from xQTL summary statistics, including joint and conditional regression analysis, Bayesian finemapping, aggregation tests, and penalized regression / variable selection.    
 
## Additional documentation pages
 See above for documentation pages on each analysis mode (cis, trans, meta, store).  Additional documentation pages are listed below. 
 - **[Input file formats for individual-level analysis](/doc/input_files)** in modes `yax cis`, `yax trans`, and `yax store`. 
 - **YAX xQTL summary association data file formats** generated by `yax store` and used in `yax meta`. *Documentation coming soon.*
 - **Bayesian finemapping using YAX xQTL summary association data files** generated by `yax store`. *Documentation coming soon.*
 
## Installation
YAX is primarily written in C++. To compile YAX from source, run:
```
git clone https://github.com/corbinq/yax.git
cd yax 
bash get_dependencies.sh
make
```
Precompiled binaries are also available for 64-bit Linux systems as follows:
```
git clone https://github.com/corbinq/yax.git
cd yax/bin
gunzip yax_Linux_x86_64.gz
mv yax_Linux_x86_64 yax && chmod +x yax
./yax --help
```

## Software dependencies

 - [Eigen C++ matrix library](http://eigen.tuxfamily.org/)
 - [HTSlib C library](http://www.htslib.org/)
 - [github.com/jonathonl/shrinkwrap](https://github.com/jonathonl/shrinkwrap)
 - [github.com/Taywee/args](https://github.com/Taywee/args)
 - [Boost math C++ library](https://www.boost.org/)

## Contact, feedback, bug reports
YAX is a work in progress, and under active development. We appreciate feedback and bug reports, and welcome any questions or feature requests. **Contact:** <qcorbin@hsph.harvard.edu>.  

## Citation
If you use YAX, please cite `Quick, C; Guan, L; Lin, X (2020). URL: https://github.com/corbinq/yax` (preprint pending). Thanks!

## Acknowledgements
 - Li Guan
 - Xihao Li
 - Zilin Li
 - Rounak Dey
 - Laura Scott
 - Hyun Min Kang 
 - Xihong Lin 


