APEX is a toolkit for analysis of molecular quantitative trait loci (xQTLs), such as mRNA expression or methylation levels. It performs cis and trans analysis, single- or multiple-variant analysis, and single-study or meta-analysis. 

We have conducted benchmarking and simulation experiments to ensure that APEX is accurate, efficient, and scalable for small and large sample sizes; [**see benchmarking**](/apex/doc/benchmarking/) for further details. 

To install APEX from source or download precompiled binaries, [**see installation guide below**](#installation).  Notable features include

**Fast, scalable linear mixed model for xQTL mapping**
 - Highly optimized linear mixed model (LMM) framework for tens of thousands of traits. This can be used to account for cryptic or known familial relatedness in cis and trans xQTL analysis. 
 - Optimized ordinary least squares (OLS) framework for cis and trans xQTL analysis with no related individuals.

**Multiple-variant and omnibus xQTL analysis**
 - Fast, rigorous analytic p-values for cis-region omnibus tests. 
 - Variable selection and conditional analysis procedures to identify multiple association signals for a single molecular trait.
 
**Flexible multiple-variant xQTL meta-analysis framework** 
 - Compact storage formats for xQTL summary statistics, including study-specific LD information. 
 - Single and multiple-variant xQTL meta-analysis procedures.
 - R interface for joint and conditional analysis, Bayesian finemapping, aggregation tests, and penalized regression from xQTL summary data files across one or multiple studies.
 
## Analysis modes and documentation
- [**xQTL factor analysis.**](/apex/doc/mode_factor/)  `./apex factor {options}`
	 - Rapid high-dimensional factor analysis to account for technical and biological variation in molecular trait data. 
- [**xQTL linear mixed models.**](/apex/doc/mode_factor/)  `./apex lmm {options}`
	 - Precompute null model terms for rapid LMM association analysis. 
- [**cis-xQTL analysis.**](/apex/doc/mode_cis/) `./apex cis {options}`
	 - Analyze xQTL associations within a genomic window of each molecular trait.  For example, analyze variant associations within 1 megabase (Mb) of each gene's transcription start site (TSS).  
 - [**trans-xQTL analysis.**](/apex/doc/mode_trans/)  `./apex trans {options}`
	 - Analyze xQTL genome-wide associations for each molecular trait, optionally excluding variants within the cis region. 
 - [**xQTL meta-analysis.**](/apex/doc/mode_meta/)  `./apex meta {options}`
	 -  Meta-analyze xQTL association summary statistics across one or multiple studies.   
 - [**xQTL summary data storage.**](/apex/doc/mode_store/)  `./apex store {options}`
	 -  Store genotypic variance-covariance (LD) data matrices for data-sharing or meta-analysis. These files enable accurate multiple-variant analysis from xQTL summary statistics, including joint and conditional regression analysis, Bayesian finemapping, aggregation tests, and penalized regression / variable selection.    
 
## Additional documentation pages
 See above for documentation pages on each analysis mode (cis, trans, meta, store).  Additional documentation pages are listed below. 
 - **[Input file formats for individual-level analysis](/apex/doc/input_files)** in modes `apex cis`, `apex trans`, and `apex store`. 
 - **APEX xQTL summary association data file formats** generated by `apex store` and used in `apex meta`. *Documentation coming soon.*
 - **Bayesian finemapping using APEX xQTL summary association data files** generated by `apex store`. *Documentation coming soon.*
 
## Installation
APEX is primarily written in C++. To compile APEX from source, run:
```
git clone https://github.com/corbinq/apex.git
cd apex 
bash get_dependencies.sh
make
```
Precompiled binaries are also available for 64-bit Linux systems as follows:
```
git clone https://github.com/corbinq/apex.git
cd apex/bin
gunzip apex_Linux_x86_64.gz
mv apex_Linux_x86_64 apex && chmod +x apex
./apex --help
```

## Software dependencies

 - [Eigen C++ matrix library](http://eigen.tuxfamily.org/)
 - [HTSlib C library](http://www.htslib.org/)
 - [github.com/jonathonl/shrinkwrap](https://github.com/jonathonl/shrinkwrap)
 - [github.com/Taywee/args](https://github.com/Taywee/args)
 - [Boost math C++ library](https://www.boost.org/)

## Contact, feedback, bug reports
APEX is a work in progress, and under active development. We appreciate feedback and bug reports, and welcome any questions or feature requests. **Contact:** <qcorbin@hsph.harvard.edu>.  

## Citation
If you use APEX, please cite `Quick, C; Guan, L; Lin, X (2020). URL: https://github.com/corbinq/apex` (preprint pending). Thanks!

## Acknowledgements
APEX is developed by Corbin Quick and Li Guan. Special thanks to 

 - Yaowu Liu
 - Xihao Li
 - Zilin Li
 - Rounak Dey
 - Laura Scott
 - Hyun Min Kang 
 - Xihong Lin 



